ddl/backfilling_scheduler.go

```diff
@@ -202,64 +231,237 @@ func (b *backfillScheduler) adjustWorkerSize() error {
 	}
-	if b.copReqSenderPool != nil {
-		b.copReqSenderPool.adjustSize(readerCnt)
-	}
 	return injectCheckBackfillWorkerNum(len(b.workers), b.tp == typeAddIndexMergeTmpWorker)
 }
 
-func (b *backfillScheduler) initCopReqSenderPool() {
-	if b.tp != typeAddIndexWorker || b.reorgInfo.ReorgMeta.ReorgTp != model.ReorgTypeLitMerge ||
-		b.copReqSenderPool != nil || len(b.workers) > 0 {
+func (b *txnBackfillScheduler) close(force bool) {
+	if b.closed {
 		return
 	}
+	close(b.taskCh)
+	if force {
+		closeBackfillWorkers(b.workers)
+	}
+	b.wg.Wait()
+	close(b.resultCh)
+	b.closed = true
+}
+
+type ingestBackfillScheduler struct {
+	ctx       context.Context
+	reorgInfo *reorgInfo
+	tbl       table.PhysicalTable
+
+	closed bool
+
+	taskCh   chan *reorgBackfillTask
+	resultCh chan *backfillResult
+
+	copReqSenderPool *copReqSenderPool
+
+	writerPool  *workerpool.WorkerPool[idxRecResult]
+	writerMaxID atomic.Int64
+	poolErr     chan error
+	engineInfo  *ingest.EngineInfo
+}
+
+func newIngestBackfillScheduler(ctx context.Context, info *reorgInfo, tbl table.PhysicalTable) *ingestBackfillScheduler {
+	return &ingestBackfillScheduler{
+		ctx:         ctx,
+		reorgInfo:   info,
+		tbl:         tbl,
+		taskCh:      make(chan *reorgBackfillTask, backfillTaskChanSize),
+		resultCh:    make(chan *backfillResult, backfillTaskChanSize),
+		writerMaxID: atomic.Int64{},
+		poolErr:     make(chan error),
+	}
+}
+
+func (b *ingestBackfillScheduler) setupWorkers() error {
+	job := b.reorgInfo.Job
+	bc, ok := ingest.LitBackCtxMgr.Load(job.ID)
+	if !ok {
+		logutil.BgLogger().Error(ingest.LitErrGetBackendFail, zap.Int64("job ID", job.ID))
+		return errors.Trace(errors.New("cannot get lightning backend"))
+	}
+	ei, err := bc.EngMgr.Register(bc, job.ID, b.reorgInfo.currElement.ID, job.SchemaName, job.TableName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	b.engineInfo = ei
+	b.copReqSenderPool, err = b.createCopReqSenderPool()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	_, writerCnt := b.expectedWorkerSize()
+	skipReg := workerpool.OptionSkipRegister[idxRecResult]{}
+	writerPool, err := workerpool.NewWorkerPool[idxRecResult]("ingest_writer",
+		poolutil.DDL, writerCnt, b.createWorker, skipReg)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	b.reorgInfo.d.setDDLLabelForTopSQL(job.ID, job.Query)
+	b.writerPool = writerPool
+	b.copReqSenderPool.resultsCh = writerPool
+	return b.adjustWorkerSize()
+}
+
+func (b *ingestBackfillScheduler) close(force bool) {
+	if b.closed {
+		return
+	}
+	close(b.taskCh)
+	b.copReqSenderPool.close(force)
+	b.writerPool.ReleaseAndWait()
+	close(b.resultCh)
+	if !force {
+		jobID := b.reorgInfo.ID
+		indexID := b.reorgInfo.currElement.ID
+		if bc, ok := ingest.LitBackCtxMgr.Load(jobID); ok {
+			bc.EngMgr.ResetWorkers(bc, jobID, indexID)
+		}
+	}
+	b.closed = true
+}
+
+func (b *ingestBackfillScheduler) sendTask(task *reorgBackfillTask) error {
+	select {
+	case b.taskCh <- task:
+		return nil
+	case err := <-b.poolErr:
```

Comment:
It makes confuse that check the error of `create worker` in sendTask.

Reply:
Otherwise, there is no way to inform the DDL worker thread to know there is an error.

Reply:
Updated.